### PR TITLE
Add migration to remove  v1 hash objects.

### LINKED ISSUE #2962: Add schema migration to remove 'hashing_version = 1' hash_objects
To address #2955  we've decided to simply delete all v1 hash objects since:

* They're causing issues here since a single v1 hash might be mapped to multiple v2 objects if the database was missing objects for branches due to unexpected sync interruptions (this error has been fixed, but base has some of these cases).
* We don't really have anyone depending on v1 hashes at the moment.

We _could_ update the v1 -> v2 migration to do this, but certain codebases like `base` have already been working on a migrated codebase for a while, so it's just as easy to add a new v2 -> v3 migration which simply deletes those hash objects.

As part of fixing this, I'll also slightly simplify the migration pattern so it's easier to add migrations and ucm will always pick the right migrations to run.

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

The 1 to 2 migration kept around hash objects of hash version 1, unfortunately this
caused an issue:

The migration would detect causals whose value hash did not have a corresponding branch
object (this was caused by a race-condition in sync which could end up in a partial sync).

When a branch object was determined to be missing, the migration replaces it with the
empty branch. This worked well, but led to a situation where related parent or successors
of that causal would have their hash objects mapped to the new v2 object which contained
the empty branch in place of missing branches. This is fine, but, if a different codebase
migrated the same branch and wasn't missing the branch in question it would migrate
successfully and each database now have the same v1 hash object mapped to two distinct v2
objects, which rightfully causes a crash when syncing.

This migration drops all the v1 hash objects to avoid this issue, since these hash objects
weren't being used for anything anyways.

## Implementation notes

* Adds a new v3 schema version
* Adds 2 to 3 migration which deletes all v1 hash objects

# Additional Notes:

See discussion here #2966 , it explores the idea of adding a stronger constraint on the causal table to prevent database integrity issues like this in the future.

# Testing

TODO:

- [x] Test this on all failures in #2955 
- [x] Test merge behaviour on two branches which differ only on a distant ancestor (since this change can result in this situation, specifically in base)
  - Succeeds, but takes a while, I suspect the most expensive part is constructing the diff. I think we should probably not compute the diff and just print the command to run in order to diff results instead.

### CODE REVIEW

**Comment by aryairani** [comment]
Probably just want to check out the trunk versions of `DbId.hs` and `Queries.hs` and reapply the few changes manually.

File: codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs [hunk 1]
```diff
@@ -873,6 +876,18 @@ getCausalsWithoutBranchObjects = queryAtoms_ sql
...
+removeHashObjectsByHashingVersion :: DB m => HashVersion -> m ()
+removeHashObjectsByHashingVersion hashVersion =
+  execute sql (Only hashVersion)
+  where
+    sql =
+      [here|
+    DELETE FROM hash_object
+      WHERE hash_version = ?
+|]
+
```
Comment (aryairani) [inline codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs:890]: Weird formatting, lol; might want to fix it up or tuck it into the {- unformatted -} section. Or leave it as-is; also fine.

**Review by aryairani — APPROVED** [review]
Great, thanks!

File: codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs (hunk 1 above)
Comment (ChrisPenner) [reply codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs:890]: Yeah it's a bit weird that ormolu forces the `sql` and the `[here|` on different lines, but I don't personally find it to hard to read, it's such a short block 🤷🏼‍♂️

**Pull request merged**